### PR TITLE
[Feature] Enable Lance scans with catalog cloud credentials

### LINKED ISSUE #74665: [Feature] Native Lance connector for StarRocks
- [x] https://github.com/StarRocks/starrocks/pull/74659
## Is your feature request related to a problem? Please describe.

StarRocks currently supports external table connectors for Iceberg, Delta Lake, Hudi, Paimon, and other formats. However, there is no support for [Lance](https://github.com/lancedb/lance), a modern columnar data format optimized for ML workloads — embedding vectors, multi-modal data, and large-scale training datasets.

Users who store datasets in Lance format must currently ETL data into a supported format before querying with StarRocks, adding latency and operational complexity.

## Describe the solution you'd like

A native Lance connector that allows StarRocks to directly query Lance datasets on local and cloud filesystems (S3, ADLS Gen2, GCS), enabling:

- Ad-hoc exploration of ML training datasets (embeddings, document corpora)
- Quality analysis by joining Lance datasets with metadata in Iceberg/Hive tables
- Interactive analytics replacing Spark read paths on Lance tables

## Implementation Plan

The implementation is split into independently reviewable PRs:

| PR | Scope | Status |
|----|-------|--------|
| #74659 | Thrift definitions for Lance scan range fields | Open |
| #74660 | FE optimizer operators and rules | Open |
| #74661 | FE scan node and plan fragment builder | Open |
| #74662 | Java-extensions Lance JNI reader | Open |
| #74663 | BE optional pluggable connector module (follows benchmark pattern) | Open |
| #74664 | Build flag `--without-connector-lance` | Open |

**Already merged (Phase 1):** FE catalog & metadata discovery (`LanceConnector`, `LanceMetadata`, `LanceApiConverter`, `LanceTable`)

## Architecture

- **FE**: `LANCE` registered as `ConnectorType` and `TableType`. Query planning reuses `THdfsScanRange` with Lance-specific fields. CBO statistics integration.
- **BE**: Optional pluggable module following the benchmark connector pattern (`--without-connector-lance` compile-time opt-out). JNI scanner delegates to `lance-core` Java SDK.
- **java-extensions**: `LanceSplitScanner` using Arrow vectors via `lance-core:7.0.0`

## Future Work

- Predicate pushdown expression visitor
- Cloud storage credential passthrough (Azure Workload Identity, S3 IAM, GCS)
- HMS-backed catalog discovery (replacing property-based bootstrap)
- Native Rust FFI scanner (zero-copy via Arrow C Data Interface)

## Are you willing to submit PR?

Yes — PRs are already submitted (see table above).

### PR DESCRIPTION
**Pull request by beinan**
## Why I'm doing:

Lance SQL execution must preserve filters and use the catalog credentials when opening private datasets.

## What I'm doing:

Support Arrow LargeUtf8/LargeBinary and widen unsigned integers without overflow (UInt8/16/32/64 to SMALLINT/INT/BIGINT/DECIMAL(20,0)); align FE schema declarations with reader decoding.

Add the optional BE connector and table descriptor dispatch, then enable the SQL relation transformer and supported-table flag after the planner, packaged reader, and BE path are all present.

Preserve WHERE predicates in Thrift and evaluate them on decoded JNI chunks. Mark JNI slots nullable and retain projected/predicate columns and LIMIT semantics. Carry the catalog identity and cloud configuration through FE planning, Thrift, BE, and Lance storage options. Support static/session S3 credentials and the native default credential chain, plus account-qualified Azure SAS/shared-key and standard ADLS2 managed-identity/service-principal configuration.

Exclude cloud properties and dataset URIs from JNI parameter logs; reader descriptions and native error handling do not expose credentials. Unsupported credential modes fail explicitly. SAS/session credentials are static; remote credential-service refresh, explicit AWS STS role assumption, and Azure workload-identity token files are outside this implementation. The English [Lance catalog guide](https://github.com/beinan/starrocks/blob/lance/pr5-be-connector/docs/en/data_source/catalog/lance_catalog.md) describes supported configuration and limitations.

Series order: #74660 → #74661 → #74662 → #74663 → #74664. Catalog and Thrift prerequisites (#74420 and #74659) are already merged. These branches are cumulative against `main`; each preceding layer is required.

### Validation

- FE production and test sources compile. All 18 focused Lance FE tests plus the statistics validity regression pass, including two-table join/aggregation, WHERE/LIMIT, projection, and credential serialization.
- All 32 reader tests pass, including native multi-fragment reads, date64, cloud option conversion, credential-safe descriptions, and an isolated classloader subprocess.
- Six affected C++ translation units compile in the upstream development image: Lance connector, JNI scanner, table descriptors, extended descriptors, execution factory, and connector bootstrap.
- FE/reader Checkstyle, BE module-boundary and generated-guide checks pass.

This validation covers FE planning and native reader execution. A full FE-to-BE SQL runtime test and live S3/Azure authentication test of this OSS revision have not been run.

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function
  - [ ] This pr needs auto generate documentation
- [ ] This is a backport pr

Documentation for user-visible execution is included in #74663; this series targets `main` only.

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.1
  - [ ] 4.0
  - [ ] 3.5

## Related Issue

Fixes #74665

This PR implements one layer of the linked feature. The complete implementation requires all five PRs in the series listed above.

### CODE REVIEW

**Comment by CelerData-Reviewer** [comment]
@codex review

File: be/src/connector/lance/lance_connector.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+    std::string scanner_factory_class = "com/starrocks/lance/reader/LanceSplitScannerFactory";
+    _scanner = std::make_unique<JniScanner>(scanner_factory_class, jni_scanner_params);
+
+    // Build minimal HdfsScannerParams needed by JniScanner
+    HdfsScannerParams scanner_params;
+    scanner_params.tuple_desc = _tuple_desc;
+    for (auto* slot : _tuple_desc->slots()) {
+        scanner_params.materialize_slots.push_back(slot);
+    }
```
[the remark this answers is not in the document]
Comment (beinan) [reply be/src/connector/lance/lance_connector.cpp]: Valid point. The current implementation populates tuple_desc, materialize_slots, and conjunct contexts. Will check if additional fields (like scan_range, partition_values) are needed for JniScanner initialization and fix if necessary.

**Comment by beinan** [comment]
@github-actions run-checks

File: be/src/connector/lance/lance_connector.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (beinan) [reply be/src/connector/lance/lance_connector.cpp]: Addressed in 0a2bf8c: LanceDataSource now populates `HdfsScannerContext` with `tuple_desc`, `materialize_slots`, `materialize_index_in_chunk`, and `scan_range` before calling `_scanner->init(state, &_scanner_ctx)`. Also switched from the removed `HdfsScannerParams` to the merged `HdfsScannerContext` struct, and the dataset URI is now read from the new `LanceTableDescriptor` so the BE no longer dereferences the removed `THdfsScanRange.lance_dataset_uri`.

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by beinan** [comment]
@dirtysalt @stdpain Conflicts have been resolved and the Lance PR stack has been rebased onto the latest main. This PR follows #74662. Could you please take another look when available? CI is rerunning on the updated commits.

File: java-extensions/lance-reader/src/main/java/com/starrocks/lance/reader/LanceColumnValue.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    @Override
+    public LocalDateTime getDateTime(ColumnType.TypeValue type) {
+        if (vector instanceof TimeStampMicroVector) {
+            long micros = ((TimeStampMicroVector) vector).get(rowIndex);
+            long epochSecond = micros / 1_000_000;
+            int nanoAdjustment = (int) ((micros % 1_000_000) * 1000);
+            return LocalDateTime.ofInstant(
+                    Instant.ofEpochSecond(epochSecond, nanoAdjustment), ZoneOffset.UTC);
+        }
+        return LocalDateTime.ofInstant(Instant.EPOCH, ZoneOffset.UTC);
```
[the remark this answers is not in the document]
Comment (beinan) [reply java-extensions/lance-reader/src/main/java/com/starrocks/lance/reader/LanceColumnValue.java]: Fixed in 3113c53f4b, inherited by this branch. LanceColumnValue now decodes SECOND, MILLISECOND, MICROSECOND, and NANOSECOND Arrow timestamp vectors correctly, with tests for all units and negative values.

**Comment by beinan** [comment]
@dirtysalt @stdpain Review feedback has been addressed in the updated stack, including invalid Lance statistics fallback in 60146b0acd. All review threads are resolved and the PR is mergeable. Please take another look when available.

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by beinan** [comment]
Rebased onto the latest main (conflict-free) and fixed both CI failures.

**FE UT (`StatisticsCalculatorTest.testValidConnectorOutputRowCount`)** — this was a genuine test bug, not flakiness. `Statistics.Builder#setOutputRowCount` clamps its argument into `[1, MAXIMUM_ROW_COUNT]` (added upstream in #67777), so `setOutputRowCount(POSITIVE_INFINITY)` becomes `MAXIMUM_ROW_COUNT` and `setOutputRowCount(-1)` becomes `1` — both finite and non-negative. The assertions expecting `false` for those two cases were therefore unreachable through the builder. The test now asserts the clamping contract instead, and keeps the case the Lance fallback actually depends on: a connector metadata that does not override `getTableStatistics` returns a builder-default `Statistics` whose row count is `NaN`, which `hasValidOutputRowCount` correctly rejects.

**BE incremental coverage (6/84)** — added `LanceTableDescriptorTest` in `be/test/runtime/descriptors_test.cpp`, covering thrift parsing, both `__isset` fallbacks that yield an empty dataset URI, and `DescriptorTbl::create` dispatch for `TTableType::LANCE_TABLE`.

The Java-Extensions coverage fix (missing `@{argLine}`, which silently disabled JaCoCo for `lance-reader`) is inherited from #74662.

**Comment by beinan** [comment]
@dirtysalt @stdpain — fourth PR in the Lance connector stack (BE connector module).

Status: **CI fully green** on the current head, rebased onto latest main, no conflicts. Both earlier failures are fixed:
- **FE UT** was a genuine test bug: `Statistics.Builder#setOutputRowCount` clamps into `[1, MAXIMUM_ROW_COUNT]` (upstream #67777), so the assertions expecting infinity/-1 to be rejected were unreachable through the builder. The test now asserts the clamping contract and keeps the case the Lance fallback actually relies on — a connector metadata that does not override `getTableStatistics` yields `NaN`, which is correctly rejected.
- **BE incremental coverage** — added `LanceTableDescriptorTest` covering thrift parsing, both `__isset` fallbacks, and `DescriptorTbl::create` dispatch for `TTableType::LANCE_TABLE`.

Your feedback on `HdfsScannerContext` population (tuple_desc, materialize_slots, materialize_index_in_chunk, scan_range) is implemented.

Merge order: #74660 → #74661 → #74662 → **#74663** → #74664.
